lib/http_aws_sigv4.c

```diff
@@ -44,67 +44,259 @@
 #include "curl_memory.h"
 #include "memdebug.h"
 
+#include "slist.h"
+
 #define HMAC_SHA256(k, kl, d, dl, o)        \
   do {                                      \
     ret = Curl_hmacit(Curl_HMAC_SHA256,     \
                       (unsigned char *)k,   \
                       (unsigned int)kl,     \
                       (unsigned char *)d,   \
                       (unsigned int)dl, o); \
-    if(ret != CURLE_OK) {                   \
+    if(ret) {                               \
       goto fail;                            \
     }                                       \
   } while(0)
 
+#define TIMESTAMP_SIZE 17
+
 static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
 {
   int i;
 
   DEBUGASSERT(dst_l >= 65);
   for(i = 0; i < 32; ++i) {
-    curl_msnprintf(dst + (i * 2), dst_l - (i * 2), "%02x", sha[i]);
+    msnprintf(dst + (i * 2), dst_l - (i * 2), "%02x", sha[i]);
+  }
+}
+
+static char *find_date_hdr(struct Curl_easy *data, const char *sig_hdr)
+{
+  char *tmp = Curl_checkheaders(data, sig_hdr, strlen(sig_hdr));
+
+  if(tmp)
+    return tmp;
+  return Curl_checkheaders(data, STRCONST("Date"));
+}
+
+/* remove whitespace, and lowercase all headers */
+static void trim_headers(struct curl_slist *head)
+{
+  struct curl_slist *l;
+  for(l = head; l; l = l->next) {
+    char *value; /* to read from */
+    char *store;
+    size_t colon = strcspn(l->data, ":");
+    Curl_strntolower(l->data, l->data, colon);
+
+    value = &l->data[colon];
+    if(!*value)
+      continue;
+    ++value;
+    store = value;
+
+    /* skip leading whitespace */
+    while(*value && ISBLANK(*value))
+      value++;
+
+    while(*value) {
+      int space = 0;
+      while(*value && ISBLANK(*value)) {
+        value++;
+        space++;
+      }
+      if(space) {
+        /* replace any number of consecutive whitespace with a single space,
+           unless at the end of the string, then nothing */
+        if(*value)
+          *store++ = ' ';
+      }
+      else
+        *store++ = *value++;
+    }
+    *store = 0; /* null terminate */
   }
 }
 
+/* maximum lenth for the aws sivg4 parts */
+#define MAX_SIGV4_LEN 64
+#define MAX_SIGV4_LEN_TXT "64"
+
+#define DATE_HDR_KEY_LEN (MAX_SIGV4_LEN + sizeof("X--Date"))
+
+/* FQDN + host: */
+#define FULL_HOST_LEN (255 + sizeof("host:"))
+
+/* string been x-PROVIDER-date:TIMESTAMP, I need +1 for ':' */
+#define DATE_FULL_HDR_LEN (DATE_HDR_KEY_LEN + TIMESTAMP_SIZE + 1)
+
+/* strlen(timestamp) need to be inferior to TIMESTAMP_SIZE */
```

Comment:
I propose:
~~~
/* sizeof(timestamp) needs to be at least TIMESTAMP_SIZE  bytes */

Reply:
maybe it's my english lacking here, but doesn't "need to be at last", mean that strlen(timestamp) must be 17 or more ? which would mean an overflow ?
I've just patch to comment to say:
```
strlen(timestamp) should be 16, and having it bigger than TIMESTAMP_SIZE could cause an overflow
```
it's more tedious, but more precise I guess.

Reply:
strlen() is the length of a string, not the size of a buffer so that is just plain wrong to describe how large a buffer is. Also, the word "inferior" does not mean shorter or smaller, it means "worse" or "less good", so it makes no sense there.

My version of the comment describes how large the buffer needs to be. It needs to be at least 17 bytes. Right?

Reply:
I've change it to that
```C
/*
 * timestamp should point to a buffer of at last TIMESTAMP_SIZE bytes,
 * if it point to a buffer bigger than TIMESTAMP_SIZE bytes
 * code could overflow
 */
```

Reply:
Explain this. How would it "overflow" if the buffer is larger than that? The code assumes 17 bytes, it does not touch the bytes beyond that does it?

Reply:
Ok, overflow was wrong sorry (I thought I use strcpy, but I use `n` functions),
```C
msnprintf(date_full_hdr, DATE_FULL_HDR_LEN,
                "x-%s-date:%s", provider1, timestamp);
```
If timestamp is too big here, the signature will fail.

on the other hand timestamp been set like this:
```C
if(!strftime(timestamp, sizeof(timestamp), "%Y%m%dT%H%M%SZ", &tm)) {
``` 

I don't think it can be something else than 17. 
But if the code calling `make_headers` is modify, then timestamp could be too large for the buffer in which it is copy.

Reply:
> If timestamp is too big here, the signature will fail.

Yes, but the comment speaks of the *buffer size* for `timestamp`, now you are talking about the string put into that buffer. They are not the same.

> if(!strftime(timestamp, sizeof(timestamp), "%Y%m%dT%H%M%SZ", &tm)) {

Right, and this code makes sure it doesn't overwrite the buffer. It is works perfectly fine with a larger one.

I claim your updated comment is wrong. The buffer needs to be *at least* TIMESTAMP_SIZE bytes big. If it is bigger, the code still works.

Reply:
ok I've update the comment, thanks for the review.